File: julia/clock/clock.jl
using Printf
using Dates
import Base: +, -, ==, show


struct Clock
    hours::Int
    minutes::Int

    function Clock(hours::Int, minutes::Int)
        d_hrs, mins = divrem(minutes, 60)
        while mins < 0
            mins += 60
            d_hrs -= 1
        end

        hrs = (hours + d_hrs) % 24
        while hrs < 0
            hrs += 24
        end

        new(hrs, mins)
    end

end


# add/subtract minutes
function +(clock::Clock, minutes::Minute)::Clock
    Clock(clock.hours, clock.minutes + minutes.value)
end

function -(clock::Clock, minutes::Minute)::Clock
    clock + -minutes
end


# equality
function ==(a::Clock, b::Clock)
    a.hours == b.hours && a.minutes == b.minutes
end


function show(io::IO, c::Clock)
    timestamp = @sprintf("%02d:%02d", c.hours, c.minutes)
    show(io, timestamp)
end

File: euphoria/clock/clock.ex
namespace clock

include std/math.e
include std/search.e

constant MINS_PER_HOUR = 60
constant MINS_PER_DAY = 24 * MINS_PER_HOUR


type clock(sequence data)
    return
        length(data) = 1 and
        integer(data[1]) and
        is_in_range(data[1], {0, MINS_PER_DAY}, "[)")
end type


public function create(integer hours, integer minutes)
    integer normalized = mod(hours * MINS_PER_HOUR + minutes, MINS_PER_DAY)
    return {normalized}
end function


public function add(clock c, integer minutes)
    return create(0, c[1] + minutes)
end function

public function subtract(clock c, integer minutes)
    return add(c, -minutes)
end function


public function toString(clock c)
    integer hour = trunc(c[1] / MINS_PER_HOUR)
    integer mins = mod(c[1], MINS_PER_HOUR)
    return sprintf("%02d:%02d", {hour, mins})
end function


public function equalClocks(clock c1, clock c2)
    return c1[1] = c2[1]
end function
